File: app/src/main/java/com/ywwynm/everythingdone/permission/PermissionUtil.java
package com.ywwynm.everythingdone.permission;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.core.content.ContextCompat;

import com.ywwynm.everythingdone.helpers.AttachmentHelper;
import com.ywwynm.everythingdone.model.Thing;

import java.util.List;

/**
 * Created by ywwynm on 2016/7/8.
 * utils for Permission.
 *
 * <p>API 33+ uses split media permissions ({@code READ_MEDIA_IMAGES},
 * {@code READ_MEDIA_VIDEO}, {@code READ_MEDIA_AUDIO}). Older code that asked for
 * "any media access" via {@link #getStoragePermissions()} should migrate to the
 * type-specific helpers below — denying audio access should not block image
 * attachment loading, and vice versa.
 */
public class PermissionUtil {

    public static String[] getImagePermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return new String[] { Manifest.permission.READ_MEDIA_IMAGES };
        }
        return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
    }

    public static String[] getVideoPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return new String[] { Manifest.permission.READ_MEDIA_VIDEO };
        }
        return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
    }

    public static String[] getAudioPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return new String[] { Manifest.permission.READ_MEDIA_AUDIO };
        }
        return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
    }

    public static boolean hasImagePermission(Context context) {
        return allGranted(context, getImagePermissions());
    }

    public static boolean hasVideoPermission(Context context) {
        return allGranted(context, getVideoPermissions());
    }

    public static boolean hasAudioPermission(Context context) {
        return allGranted(context, getAudioPermissions());
    }

    /**
     * Returns only the permissions actually needed for the given things'
     * attachments. On Android 12 and below returns
     * {@code READ_EXTERNAL_STORAGE}; on Android 13+ returns the subset of
     * {@code READ_MEDIA_IMAGES}, {@code READ_MEDIA_VIDEO}, and
     * {@code READ_MEDIA_AUDIO} that matches the attachment types present.
     */
    public static String[] getRequiredPermissionsForThings(List<Thing> things) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
        }

        boolean needsImage = false, needsVideo = false, needsAudio = false;
        for (Thing thing : things) {
            String attachment = thing.getAttachment();
            if (!AttachmentHelper.isValidForm(attachment)) continue;
            String[] parts = attachment.split(AttachmentHelper.SIGNAL);
            for (int i = 1; i < parts.length; i++) {
                char type = parts[i].charAt(0);
                if (type == '0') needsImage = true;
                else if (type == '1') needsVideo = true;
                else if (type == '2') needsAudio = true;
            }
        }

        java.util.List<String> perms = new java.util.ArrayList<>(3);
        if (needsImage) perms.add(Manifest.permission.READ_MEDIA_IMAGES);
        if (needsVideo) perms.add(Manifest.permission.READ_MEDIA_VIDEO);
        if (needsAudio) perms.add(Manifest.permission.READ_MEDIA_AUDIO);

        if (perms.isEmpty()) {
            perms.add(Manifest.permission.READ_MEDIA_IMAGES);
        }
        return perms.toArray(new String[0]);
    }

    /**
     * Returns the union of all media read permissions appropriate for the
     * running platform.
     *
     * @deprecated Asking for image, video and audio permissions together is
     * intrusive on Android 13+ and a denial on any one type currently aborts
     * unrelated flows. Use {@link #getImagePermissions()},
     * {@link #getVideoPermissions()} or {@link #getAudioPermissions()} for
     * new code, and consider {@code ActivityResultContracts.PickVisualMedia} or
     * SAF ({@code ACTION_OPEN_DOCUMENT}) which need no permission at all.
     */
    @Deprecated
    public static String[] getStoragePermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return new String[] {
                    Manifest.permission.READ_MEDIA_IMAGES,
                    Manifest.permission.READ_MEDIA_VIDEO,
                    Manifest.permission.READ_MEDIA_AUDIO
            };
        }
        return new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
    }

    /**
     * Returns true when the app holds at least one media read permission. Older
     * call sites used to require all three on API 33+, which meant denying audio
     * access also broke image-attachment loading. The relaxed check is enough
     * for code paths that load whichever attachments happen to exist.
     */
    public static boolean hasStoragePermission(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return hasImagePermission(context)
                    || hasVideoPermission(context)
                    || hasAudioPermission(context);
        }
        return ContextCompat.checkSelfPermission(context,
                Manifest.permission.READ_EXTERNAL_STORAGE)
                == PackageManager.PERMISSION_GRANTED;
    }

    public static boolean shouldRequestPermissionWhenLoadingThings(List<Thing> things) {
        for (Thing thing : things) {
            if (AttachmentHelper.isValidForm(thing.getAttachment())) {
                return true;
            }
        }
        return false;
    }

    private static boolean allGranted(Context context, String[] permissions) {
        for (String perm : permissions) {
            if (ContextCompat.checkSelfPermission(context, perm)
                    != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/permission/PermissionUtil.kt
package com.ywwynm.everythingdone.permission

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat
import com.ywwynm.everythingdone.helpers.AttachmentHelper
import com.ywwynm.everythingdone.model.Thing
import java.util.ArrayList

/**
 * Created by ywwynm on 2016/7/8.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * utils for Permission.
 *
 *
 * API 33+ uses split media permissions (`READ_MEDIA_IMAGES`,
 * `READ_MEDIA_VIDEO`, `READ_MEDIA_AUDIO`). Older code that asked for
 * "any media access" via [getStoragePermissions] should migrate to the
 * type-specific helpers below — denying audio access should not block image
 * attachment loading, and vice versa.
 */
object PermissionUtil {

    @JvmStatic
    fun getImagePermissions(): Array<String?>? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return arrayOf(Manifest.permission.READ_MEDIA_IMAGES)
        }
        return arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    @JvmStatic
    fun getVideoPermissions(): Array<String?>? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return arrayOf(Manifest.permission.READ_MEDIA_VIDEO)
        }
        return arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    @JvmStatic
    fun getAudioPermissions(): Array<String?>? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return arrayOf(Manifest.permission.READ_MEDIA_AUDIO)
        }
        return arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    @JvmStatic
    fun hasImagePermission(context: Context?): Boolean {
        return allGranted(context, getImagePermissions())
    }

    @JvmStatic
    fun hasVideoPermission(context: Context?): Boolean {
        return allGranted(context, getVideoPermissions())
    }

    @JvmStatic
    fun hasAudioPermission(context: Context?): Boolean {
        return allGranted(context, getAudioPermissions())
    }

    /**
     * Returns only the permissions actually needed for the given things'
     * attachments. On Android 12 and below returns
     * `READ_EXTERNAL_STORAGE`; on Android 13+ returns the subset of
     * `READ_MEDIA_IMAGES`, `READ_MEDIA_VIDEO`, and
     * `READ_MEDIA_AUDIO` that matches the attachment types present.
     */
    @JvmStatic
    fun getRequiredPermissionsForThings(things: List<Thing?>?): Array<String?>? {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        }

        var needsImage = false
        var needsVideo = false
        var needsAudio = false
        for (thing in things!!) {
            val attachment: String? = thing!!.attachment
            if (!AttachmentHelper.isValidForm(attachment)) continue
            val parts = attachment!!.split(AttachmentHelper.SIGNAL!!.toRegex()).toTypedArray()
            for (i in 1 until parts.size) {
                val type = parts[i][0]
                when (type) {
                    '0' -> needsImage = true
                    '1' -> needsVideo = true
                    '2' -> needsAudio = true
                }
            }
        }

        val perms = ArrayList<String?>(3)
        if (needsImage) perms.add(Manifest.permission.READ_MEDIA_IMAGES)
        if (needsVideo) perms.add(Manifest.permission.READ_MEDIA_VIDEO)
        if (needsAudio) perms.add(Manifest.permission.READ_MEDIA_AUDIO)

        if (perms.isEmpty()) {
            perms.add(Manifest.permission.READ_MEDIA_IMAGES)
        }
        return perms.toTypedArray()
    }

    /**
     * Returns the union of all media read permissions appropriate for the
     * running platform.
     */
    @Deprecated(
        "Asking for image, video and audio permissions together is intrusive on Android 13+ and a denial on any one type currently aborts unrelated flows. Use getImagePermissions, getVideoPermissions or getAudioPermissions for new code, and consider ActivityResultContracts.PickVisualMedia or SAF (ACTION_OPEN_DOCUMENT) which need no permission at all."
    )
    @JvmStatic
    fun getStoragePermissions(): Array<String?>? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return arrayOf(
                    Manifest.permission.READ_MEDIA_IMAGES,
                    Manifest.permission.READ_MEDIA_VIDEO,
                    Manifest.permission.READ_MEDIA_AUDIO
            )
        }
        return arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    /**
     * Returns true when the app holds at least one media read permission. Older
     * call sites used to require all three on API 33+, which meant denying audio
     * access also broke image-attachment loading. The relaxed check is enough
     * for code paths that load whichever attachments happen to exist.
     */
    @JvmStatic
    fun hasStoragePermission(context: Context?): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return hasImagePermission(context)
                    || hasVideoPermission(context)
                    || hasAudioPermission(context)
        }
        return ContextCompat.checkSelfPermission(context!!,
                Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
    }

    @JvmStatic
    fun shouldRequestPermissionWhenLoadingThings(things: List<Thing?>?): Boolean {
        for (thing in things!!) {
            if (AttachmentHelper.isValidForm(thing!!.attachment)) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    private fun allGranted(context: Context?, permissions: Array<String?>?): Boolean {
        for (perm in permissions!!) {
            if (ContextCompat.checkSelfPermission(context!!, perm!!) != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }
}
